---
jupyter:
  jupytext:
    formats: ipynb,Rmd
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.14.4
  kernelspec:
    display_name: R
    language: R
    name: ir
---

# Setup

```{r}
source("helpers.R")
```

```{r}
invisible(sapply(c(
  "TFBSTools",
  "stringr",
  # "patchwork",
  # "ggplot2",
  "Seurat",
  "GenomicRanges",
  "BSgenome.Mmusculus.UCSC.mm10",
  "JASPAR2020",
  "BRGenomics",
  "motifmatchr",
  "Signac"
), FUN = function(x) {
  suppressPackageStartupMessages(library(x, character.only = T))
}))
```

```{r}
str_glue("- project_path: {project_path}")
str_glue("- output_dir: {output_dir}")
str_glue("- plot_dir: {plot_dir}")
str_glue("- rds_path: {rds_path}")
str_glue("- raw_dir: {raw_dir}")
str_glue("- h5_path: {h5_path}")
str_glue("- fragments_path: {fragments_path}")
str_glue("- seed: {seed}")
```

```{r}
inputs <- c(
  "h5" = h5_path,
  "fragments" = fragments_path,
  "TSS_reference" = str_c(external_dir, "GREATv4.genes.mm10.tsv")
)
outputs <- c()
inputs
outputs
```

# Load data

```{r}
h5 <- Read10X_h5(inputs[["h5"]])
```

```{r}
h5 %>% names
```

```{r}
sobj <- CreateSeuratObject(
  counts = h5$`Gene Expression`,
  assay = "RNA"
)
```

```{r}
sobj
```

```{r}
sobj[["percent.mt"]] <- PercentageFeatureSet(sobj, pattern = "^MT")
```

```{r}
h5$Peaks %>% dim
h5$Peaks[str_detect(rownames(h5$Peaks), pattern = "chr"), ] %>% dim
```

```{r}
h5$Peaks <- h5$Peaks[str_detect(rownames(h5$Peaks), pattern = "chr"), ]
```

```{r tags=c()}
annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79, verbose = F)
seqlevelsStyle(annotation) <- "UCSC"
```

```{r}
sobj[["ATAC"]] <- CreateChromatinAssay(
  counts = h5$Peaks,
  sep = c(":", "-"),
  genome = "mm10",
  fragments = inputs[["fragments"]],
  annotation = annotation,
  verbose = F
)
```

# Subset to only Immature cells

```{r}
immature_cells <- (sobj[[]] %>% rownames())[sobj[[]] %>%
  rownames() %>%
  str_sub(start = -1) %>%
  str_detect("2") %>%
  which()]
immature_cells %>% length()
```

```{r}
sobj <- sobj %>% subset(cells = immature_cells)
```

## Calculate more metadata

```{r}
DefaultAssay(sobj) <- "ATAC"
genome(sobj) <- "mm10"
```

```{r}
sobj <- NucleosomeSignal(sobj)
```

```{r}
sobj <- TSSEnrichment(sobj)
```

# Find thresholds for feature filtering

```{r}
features <- c(
  "nCount_ATAC",
  "nFeature_ATAC",
  "nCount_RNA",
  "nFeature_RNA",
  "nucleosome_signal",
  "TSS.enrichment"
)

thresholds <- c()

for (i in seq_len(length(features))) {
  thresholds[[features[i]]] <- c("low" = 0, "high" = Inf)
}
thresholds
```

## Plotting functions

```{r}
plot_vln <- \(plot) plot + NoLegend() + theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank())
plot_lines <- \(plot, feat) (plot + geom_hline(yintercept = thresholds[[feat]][["high"]], linetype = "dashed") + geom_hline(yintercept = thresholds[[feat]][["low"]], linetype = "dashed")) %>% plot_vln
```

## First feature to filter:

```{r}
feat <- features[1]
VlnPlot(sobj, features = feat) %>% plot_vln %>% show()
VlnPlot(sobj, features = feat, log = T) %>% plot_vln %>% show()
```

```{r}
thresholds[[feat]][["low"]] <- 1e3
thresholds[[feat]][["high"]] <- 5e4
```

```{r}
VlnPlot(sobj, features = feat) %>% plot_lines(feat) %>% show()
VlnPlot(sobj, features = feat, log = T) %>% plot_lines(feat) %>% show()
```

## Next feature (2) to filter:

```{r}
feat <- features[2]
VlnPlot(sobj, features = feat) %>% plot_vln %>% show()
VlnPlot(sobj, features = feat, log = T) %>% plot_vln %>% show()
```

```{r}
thresholds[[feat]][["low"]] <- .8e3
thresholds[[feat]][["high"]] <- 20e3
```

```{r}
VlnPlot(sobj, features = feat) %>% plot_lines(feat) %>% show()
VlnPlot(sobj, features = feat, log = T) %>% plot_lines(feat) %>% show()
```

## Next feature (3) to filter:

```{r}
feat <- features[3]
VlnPlot(sobj, features = feat) %>% plot_vln %>% show()
VlnPlot(sobj, features = feat, log = T) %>% plot_vln %>% show()
```

```{r}
thresholds[[feat]][["high"]] <- 18e3
thresholds[[feat]][["low"]] <- 1e3
```

```{r}
VlnPlot(sobj, features = feat) %>% plot_lines(feat) %>% show()
VlnPlot(sobj, features = feat, log = T) %>% plot_lines(feat) %>% show()
```

## Next feature (4) to filter:

```{r}
feat <- features[4]
VlnPlot(sobj, features = feat) %>% plot_vln %>% show()
VlnPlot(sobj, features = feat, log = T) %>% plot_vln %>% show()
```

```{r}
thresholds[[feat]][["high"]] <- 5e3
thresholds[[feat]][["low"]] <- .8e3
```

```{r}
VlnPlot(sobj, features = feat) %>% plot_lines(feat) %>% show()
VlnPlot(sobj, features = feat, log = T) %>% plot_lines(feat) %>% show()
```

## Next feature (5) to filter:

```{r}
feat <- features[5]
VlnPlot(sobj, features = feat) %>% plot_vln %>% show()
```

```{r}
thresholds[[feat]][["high"]] <- 1.4
thresholds[[feat]][["low"]] <- .6
```

```{r}
VlnPlot(sobj, features = feat) %>% plot_lines(feat) %>% show()
```

## Next feature (6) to filter:

```{r}
feat <- features[6]
VlnPlot(sobj, features = feat) %>% plot_vln %>% show()
VlnPlot(sobj, features = feat, log = T) %>% plot_vln %>% show()
```

```{r}
thresholds[[feat]][["high"]] <- 9
thresholds[[feat]][["low"]] <- 4
```

```{r}
VlnPlot(sobj, features = feat) %>% plot_lines(feat) %>% show()
VlnPlot(sobj, features = feat, log = T) %>% plot_lines(feat) %>% show()
```

## Resulting thresholds:

```{r}
thresholds
```

# Subset:

```{r}
sobj <- subset(
  x = sobj,
  subset = (
    nCount_ATAC < thresholds[["nCount_ATAC"]][["high"]] &
      nCount_ATAC > thresholds[["nCount_ATAC"]][["low"]] &
      nFeature_ATAC < thresholds[["nFeature_ATAC"]][["high"]] &
      nFeature_ATAC > thresholds[["nFeature_ATAC"]][["low"]] &
      nCount_RNA < thresholds[["nCount_RNA"]][["high"]] &
      nCount_RNA > thresholds[["nCount_RNA"]][["low"]] &
      nFeature_RNA < thresholds[["nFeature_RNA"]][["high"]] &
      nFeature_RNA > thresholds[["nFeature_RNA"]][["low"]] &
      nucleosome_signal < thresholds[["nucleosome_signal"]][["high"]] &
      nucleosome_signal > thresholds[["nucleosome_signal"]][["low"]] &
      TSS.enrichment < thresholds[["TSS.enrichment"]][["high"]] &
      TSS.enrichment > thresholds[["TSS.enrichment"]][["low"]]
  )
)
```

# Look at subsetting result:

```{r}
for (feat in features) {
  VlnPlot(sobj, features = feat) %>%
    plot_vln %>%
    show()
  VlnPlot(sobj, features = feat, log = T) %>%
    plot_vln %>%
    show()
}
```

# Add motifs

```{r}
DefaultAssay(sobj) <- "ATAC"
Idents(sobj) <- "origin"
```

```{r}
df <- read.delim(inputs[["TSS_reference"]],
  sep = "\t", check.names = F
)
```

```{r tags=c()}
df
```

```{r tags=c()}
df$CHROM %>% unique
```

```{r}
std_chroms <- str_c("chr", c(1:19, "X", "Y", "M"))
std_chroms
```

```{r}
tss <- GRanges(
  seqnames = df$CHROM, IRanges(
    seqnames = df$CHROM, start = df$POS,
    end = df$POS, strand = df$STRAND
  ),
  symbol = df$Symbol, strand = df$STRAND, ID = df[["#Gene_IDs"]]
)
genome(tss) <- "mm10"
```

```{r}
tss %>% list
```

```{r}
gr <- granges(sobj)
# seq_keep <- seqnames(gr) %in% seqnames(BSgenome.Mmusculus.UCSC.mm10)
seq_keep <- seqnames(gr) %in% std_chroms
seq_keep <- as.vector(seq_keep)
feat_keep <- GRangesToString(grange = gr[seq_keep])
```

```{r}
sobj[["ATAC"]] <- subset(sobj[["ATAC"]], features = feat_keep)
gr <- gr[seq_keep]
genome(gr) <- "mm10"
seqlevels(gr) <- seqlevelsInUse(granges(sobj))
```

## Split peaks into proximal and distal (i.e. non-proximal)

```{r}
genome(sobj@assays$ATAC) <- "mm10"
```

```{r}
promoters <- promoters(tss)
overlaps <- findOverlaps(gr, promoters)
peak_idx_proximal <- unique(overlaps@from)

sobj@assays$ATAC@meta.features$peak_type <- "unclassified"
sobj@assays$ATAC@meta.features[peak_idx_proximal, "peak_type"] <- "proximal"
sobj@assays$ATAC@meta.features[
  sobj@assays$ATAC@meta.features$peak_type == "unclassified",
  "peak_type"
] <- "distal"

proximal_peaks <-
  rownames(sobj@assays$ATAC@meta.features[
    sobj@assays$ATAC@meta.features$peak_type
    == "proximal",
  ])
distal_peaks <-
  rownames(sobj@assays$ATAC@meta.features[
    sobj@assays$ATAC@meta.features$peak_type
    == "distal",
  ])

sobj[["proximal"]] <- subset(sobj[["ATAC"]], features = proximal_peaks)
sobj[["distal"]] <- subset(sobj[["ATAC"]], features = distal_peaks)
```

```{r}
pfm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(
    collection = "CORE",
    tax_group = "vertebrates",
    all_versions = F
  )
)
```

```{r}
DefaultAssay(sobj) <- "proximal"
sobj <- AddMotifs(
  object = sobj,
  genome = BSgenome.Mmusculus.UCSC.mm10,
  assay = "proximal",
  pfm = pfm
)
```

```{r}
sobj <- AddMotifs(
  object = sobj,
  genome = BSgenome.Mmusculus.UCSC.mm10,
  assay = "distal",
  pfm = pfm
)
```

# Save Seurat object

```{r}
saveRDS(sobj, file = str_glue("{output_dir}seurat_object_preprocessed.rds"))
```
